Sources/GRPC/GRPCKeepaliveHandlers.swift

```diff
@@ -0,0 +1,297 @@
+    context.eventLoop.scheduleTask(in: delay) { [unowned self] in
+      self.perform(action: self.pingHandler.pingFired(), context: context)
+      self.scheduleClose(timeout: timeout, context: context)
+    }
+  }
+
+  private func scheduleClose(timeout: TimeAmount, context: ChannelHandlerContext) {
+    self.scheduledClose = context.eventLoop.scheduleTask(in: timeout) { [unowned self] in
+      self.close(context: context)
+    }
+  }
+
+  private func cancelScheduledTimeout() {
+    self.scheduledClose?.cancel()
+    self.scheduledClose = nil
+  }
+}
+
+final class PingHandler {
+  /// Code for ping
+  private let pingCode: UInt64
+
+  /// The amount of time to wait before sending a keepalive ping.
+  private let interval: TimeAmount
+
+  /// The amount of time to wait for an acknowledgment.
+  /// If it does not receive an acknowledgment within this time, it will close the connection
+  private let timeout: TimeAmount
+
+  /// Send keepalive pings even if there are no calls in flight.
+  private let permitWithoutCalls: Bool
+
+  /// Maximum number of pings that can be sent when there is no data/header frame to be sent.
+  private let maximumPingsWithoutData: UInt
+
+  /// If there are no data/header frames being received:
+  /// The minimum amount of time to wait between successive pings.
+  private let minimumSentPingIntervalWithoutData: TimeAmount
+
+  /// If there are no data/header frames being sent:
+  /// The minimum amount of time expected between receiving successive pings.
+  /// If the time between successive pings is less than this value, then the ping will be considered a bad ping from the peer.
+  /// Such a ping counts as a "ping strike".
+  private let minimumReceivedPingIntervalWithoutData: TimeAmount?
+
+  /// Maximum number of bad pings that the server will tolerate before sending an HTTP2 GOAWAY frame and closing the connection.
+  /// Setting it to `0` allows the server to accept any number of bad pings.
+  private let maximumPingStrikes: UInt?
+
+  /// When the last ping was received
+  private var lastReceivedPingDate: NIODeadline? = nil
+
+  /// When the last ping was sent
+  private var lastSentPingDate: NIODeadline? = nil
+
+  /// The number of pings sent on the transport without any data
+  private var sentPingsWithoutData = 0
+
+  /// Number of strikes
+  private var pingStrikes: UInt = 0
+
+  /// The scheduled task which will close the connection.
+  private var scheduledClose: Scheduled<Void>? = nil
+
+  /// Number of active streams
+  private var activeStreams = 0 {
+    didSet {
+      if activeStreams > 0 {
+        self.sentPingsWithoutData = 0
+      }
+    }
+  }
+
+  private let goAwayFrame = HTTP2Frame.FramePayload.goAway(lastStreamID: .rootStream, errorCode: .enhanceYourCalm, opaqueData: nil)
+
+  enum Action {
+    case schedulePing(delay: TimeAmount, timeout: TimeAmount)
+    case cancelScheduledTimeout
+    case reply(HTTP2Frame.FramePayload)
+  }
+
+  init(
```

Comment:
The testability of this is really good at the moment, except for having to deal with time. Passing in a mechanism to provide `NIODeadline`s would allow you to control time without `sleep`ing in tests.

Reply:
Agreed, but I'm just not sure how to implement that.

Reply:
I tried something, let me know if that is good enough. (Pretty much copied what was in `GRPCTimeout`).